vault/plugin_catalog.go

```diff
@@ -325,43 +375,85 @@ func (c *PluginCatalog) getPluginTypeFromUnknown(ctx context.Context, logger log
 	}
 	merr = multierror.Append(merr, err)
 
+	pluginType, err := c.getBackendPluginType(ctx, plugin)
+	if err == nil {
+		return pluginType, nil
+	}
+	merr = multierror.Append(merr, err)
+
+	return consts.PluginTypeUnknown, merr
+}
+
+// getBackendPluginType returns an error if the plugin is not a backend plugin.
+func (c *PluginCatalog) getBackendPluginType(ctx context.Context, pluginRunner *pluginutil.PluginRunner) (consts.PluginType, error) {
+	merr := &multierror.Error{}
 	// Attempt to run as backend plugin
-	client, err := backendplugin.NewPluginClient(ctx, nil, plugin, log.NewNullLogger(), true)
+	config := pluginutil.PluginClientConfig{
+		Name:            pluginRunner.Name,
+		PluginSets:      backendpluginv5.PluginSet,
+		HandshakeConfig: backendplugin.HandshakeConfig,
+		Logger:          log.NewNullLogger(),
+		IsMetadataMode:  false,
+		AutoMTLS:        true,
+	}
+
+	var client logical.Backend
+	// Attempt to run as backend V5 plugin
+	c.logger.Debug("attempting to load credential backend plugin", "name", pluginRunner.Name)
+	pc, err := c.newPluginClient(ctx, pluginRunner, config)
 	if err == nil {
-		err := client.Setup(ctx, &logical.BackendConfig{})
+		// dispense the plugin so we can get its type
+		client, err = backendpluginv5.Dispense(ctx, pc.ClientProtocol, pc)
+		if err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("failed to load plugin as backend v5: %w", err))
+		} else {
+			c.logger.Debug("successfully dispensed v5 backend plugin", "name", pluginRunner.Name)
+			defer c.cleanupExternalPlugin(pluginRunner.Name, pc.id)
+		}
```

Comment:
Wondering if `defer client.Cleanup(ctx)` is intentionally not called for v5?

Reply:
Yes, this is intentional. If we call `client.Cleanup` for v5 we will get a deadlock here. This is because the [cleanupFunc](https://github.com/hashicorp/vault/blob/03e4cdc6978282d1659386e961aac6110bdb34f5/vault/plugin_catalog.go#L283-L286) which gets called by `client.Cleanup` holds a lock on the PluginCatalog. The PluginCatalog lock is held when `getPluginTypeFromUnknown` gets called (indirectly) from [(*PluginCatalog).Set](https://github.com/hashicorp/vault/blob/03e4cdc6978282d1659386e961aac6110bdb34f5/vault/plugin_catalog.go#L633-L634) and [(*PluginCatalog).UpgradePlugins](https://github.com/hashicorp/vault/blob/03e4cdc6978282d1659386e961aac6110bdb34f5/vault/plugin_catalog.go#L502-L503).

This is only a concern on plugin registration.